chat.js

```diff
@@ -350,6 +353,38 @@ class CommandContext {
 		}
 		return true;
 	}
+	checkR9k(user, room, message) {
+		let checksum = Chat.md5(toId(message)).substr(0, 16);
+
+		if (!room.r9kTempCache) room.r9kTempCache = new Set();
+
+		if (room.r9kTempCache.has(checksum) || (room.r9kPermanentCache && room.r9kPermanentCache.has(checksum))) {
+			// If the phrase is in the temporary cache, add it to the permanent cache
+			if (!room.r9kPermanentCache) room.r9kPermanentCache = new Set();
+			if (!room.r9kPermanentCache.has(checksum)) {
+				Chat.r9kAppendPermanent(room, checksum);
```

Comment:
Mostly RAM usage. If you're adding everything eventually you'll end up with hundreds of megabytes of data per room. This way the common phrases still get banned, but most of the "noise" will not be there